feat(MMENG-4168): add mrrc-collect-params task

ligangty opened the pull request:
This commit adds a task for collecting params for MRRC from data json of collect-data task. This is part of the whole tasks for the whole Konflux MRRC Release pipeline.

mmalina commented on tasks/mrrc-collect-params/README.md:
```diff
@@ -0,0 +1,11 @@
+# mrrc-collect-params
```
What is MRRC? At the very least it should be defined in the description of the task. I even opened the Jira and that doesn't define it either.

ligangty replied:
MRRC is maven.repository.redhat.com, which is used to host maven artifacts of RedHat Middleware products. This task is one of the whole tasks for the new process to publish content to MRRC.

ligangty replied:
I'll add a simple description of the MRRC in README

mmalina replied:
Cool, thanks.

ligangty replied:
Fixed in new commit

mmalina commented on tasks/mrrc-collect-params/mrrc-collect-params.yaml:
```diff
@@ -0,0 +1,60 @@
+        echo -n "$currentPath/$mrrcEnvFile" > "$(results.mrrcParamFilePath.path)"
+      workingDir: $(workspaces.data.path)
+  workspaces:
+    - name: data
```
Missing new line at the end of the file.

ligangty replied:
Sure. Will fix.

ligangty replied:
Fixed in new commit

mmalina commented on tasks/mrrc-collect-params/README.md:
For consistency with the other tasks, I would suggest to call this `collect-mrrc-params`.

ligangty replied:
Ok. Will fix.

ligangty replied:
Fixed in new commit

mmalina commented on tasks/mrrc-collect-params/mrrc-collect-params.yaml:
```diff
@@ -0,0 +1,60 @@
+        set -eux
+
+        cd $(workspaces.data.path)
+        mrrcEnvFile="./mrrc.env"
```
I would suggest to use the same subdirectory as the data file is in. Similar to how we do it here: https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/create-pyxis-image/create-pyxis-image.yaml#L94

We store all of the data for the pipeline in a subdirectory. In the past this was needed because all of the pipeline runs were sharing a single PVC and so PLRs running in parallel would otherwise interfere.

We now have a new PVC for each PLR, so this is no longer a real issue, but it's better to be consistent. Eventually we will probably get rid of the subdirectories.

ligangty replied:
Oh, ok. Will try to use this.

ligangty replied:
Hi @mmalina 

I've tried to use this subdirectory way in my personal workspace, but it seems not working. What I used is like this:

``` yaml
apiVersion: tekton.dev/v1beta1
kind: Task
metadata:
  name: collect-mrrc-params
spec:
  params:
    - name: dataJsonPath
      type: string
      description: path to data json file
    - name: snapshotPath
      type: string
      description: Path to the JSON string of the mapped Snapshot spec in the data workspace
  results:
    - name: mrrcParamFilePath
      description: path of the env file for following tasks
    - name: cosignPubKeySecret
      description: the secret name for cosign verify public key
    - name: charonAWSSecret
      description: the secret name for charon aws credential file
  steps:
    - name: collect-mrrc-params
      image: quay.io/konflux-ci/release-service-utils:e633d51cd41d73e4b3310face21bb980af7a662f
      script: |
        set -eux

        cd $(workspaces.data.path)
        MRRC_ENV_FILE_PATH="$(dirname $(params.snapshotPath))/mrrc.env"

        zipRegistry=$(jq -r '.zipRegistry' $(params.dataJsonPath))
        echo "export MRRC_ZIP_REGISTRY=$zipRegistry" >> $MRRC_ENV_FILE_PATH

        environment=$(jq -r '.mrrc.environment' $(params.dataJsonPath))
        release=$(jq -r '.mrrc.release' $(params.dataJsonPath))
        target=$environment-maven-$release
        echo "export MRRC_TARGET=$target" >> $MRRC_ENV_FILE_PATH

        productName=$(jq -r '.mrrc.product.name' $(params.dataJsonPath))
        productVersion=$(jq -r '.mrrc.product.version' $(params.dataJsonPath))
        echo "export MRRC_PRODUCT_NAME=$productName" >> $MRRC_ENV_FILE_PATH
        echo "export MRRC_PRODUCT_VERSION=$productVersion" >> $MRRC_ENV_FILE_PATH

        awsSecret=$(jq -r '.mrrc.awsSecret' $(params.dataJsonPath))
        echo -n "$awsSecret" > "$(results.charonAWSSecret.path)"

        cosignPubKeySec=$(jq -r '.cosignPubKeySecret' $(params.dataJsonPath))
        if [ -n "$cosignPubKeySec" ] && [ "$cosignPubKeySec" != "null" ]; then 
          echo -n "$cosignPubKeySec" > "$(results.cosignPubKeySecret.path)"
        fi

        echo -n "$MRRC_ENV_FILE_PATH" > "$(results.mrrcParamFilePath.path)"
      workingDir: $(workspaces.data.path)
  workspaces:
    - name: data
```

And I passed the param by this way:
``` yaml
......
    params:
        - name: dataJsonPath
          value: "$(workspaces.data.path)/$(tasks.collect-data.results.data)"
        - name: snapshotPath
          value: "$(tasks.collect-data.results.snapshotSpec)"
......

``` 

But the output of the testing is like following:
```
+ cd /workspace/data
++ dirname /snapshot_spec.json
+ MRRC_ENV_FILE_PATH=//mrrc.env
++ jq -r .zipRegistry /workspace/data//data.json
+ zipRegistry=quay.io/ligangty/activemq.zip:0.0.1
+ echo 'export MRRC_ZIP_REGISTRY=quay.io/ligangty/activemq.zip:0.0.1'
.......
```

Seems the "$(tasks.collect-data.results.snapshotSpec)" which is from the collect-data task passed an empty value there. Is something I'm missing for the usage of this?

johnbieren replied:
Why use the snapshotSpec instead of the data file which is already a parameter to this task?

ligangty replied:
@johnbieren I'm following @mmalina 's advice to use that "snapshotPath". I just copied an example usage which assign the value to this snapshotPath: https://github.com/konflux-ci/release-service-catalog/blob/development/pipelines/fbc-release/fbc-release.yaml#L138

ligangty replied:
> Seems the "$(tasks.collect-data.results.snapshotSpec)" which is from the collect-data task passed an empty value there. Is something I'm missing for the usage of this?

Ok. After checking seems I missed to set the subdirectory param value for the collect-data. After fixing it can pass my local test.

ligangty replied:
> Why use the snapshotSpec instead of the data file which is already a parameter to this task?

@johnbieren Oh ok I know what you mean. Will use the data path directly. Thanks.

ligangty replied:
Fixed in new commit with a subdirectory setting with the dataJsonPath

mmalina replied:
Yeah, what I meant is to use the same approach as in that other task, but here we can use the data path instead of the snapshot spec path.

mmalina commented on tasks/mrrc-collect-params/mrrc-collect-params.yaml:
```diff
@@ -0,0 +1,60 @@
+
+        cd $(workspaces.data.path)
+        mrrcEnvFile="./mrrc.env"
+        if [ -f $mrrcEnvFile ]; then
```
Why is this needed? I would suggest to remove it.

ligangty replied:
Ok. Will fix.

ligangty replied:
Fixed in new commit

mmalina commented on tasks/mrrc-collect-params/mrrc-collect-params.yaml:
```diff
@@ -0,0 +1,60 @@
+        awsSecret=$(yq -r '.mrrc.awsSecret' $(params.dataJsonPath))
+        echo -n "$awsSecret" > "$(results.charonAWSSecret.path)"
+
+        consignPubKey=$(yq -r '.cosignPubKeyFile' $(params.dataJsonPath))
```
We already have `.sign.cosignSecretName` elsewhere - for cosign signing of images. I assume this will be a different secret. So maybe better to have it under `.mrrc` to make it clearer what it's related to?

Also, what is the key file name for? Is that something that can be different every time? Isn't it always the same?

ligangty replied:
This is used to provide the pub key for cosign to verify the maven zip which will be hosted in quay.io as OCI-format. The pub key will highly possible be different for different MW teams, so the basic idea is storing the pub key through configmap(or secret) and then use in the following tasks. This means it will be mount as volume. 
And the file name is the key in this configmap or secret which is the file name of that pub key file(for example, cosign.pub). It will be mounted into followig tasks to be used to do verification.

As described, this is just a basic idea yet, and may be changed later if any better solutions come out for this cosign verification task.

mmalina replied:
Thanks for the explanation. Yeah, it makes sense. Since only public keys will be involved here, it can probably be a configMap (doesn't have to be secret). But I guess it doesn't make much difference - either way it will be mounted as a volume.

Regarding the key name, I wonder if it would make sense to stick to just one name (like cosign.pub) - there is probably no benefit in having a custom name, right? I mean it shouldn't even matter which file the key value is saved to - as long as the file is passed as a parameter to whatever needs the key file.

ligangty replied:
> Thanks for the explanation. Yeah, it makes sense. Since only public keys will be involved here, it can probably be a configMap (doesn't have to be secret). But I guess it doesn't make much difference - either way it will be mounted as a volume.

Yes the configmap is ok. This will be used later in the following tasks

> Regarding the key name, I wonder if it would make sense to stick to just one name (like cosign.pub) - there is probably no benefit in having a custom name, right? I mean it shouldn't even matter which file the key value is saved to - as long as the file is passed as a parameter to whatever needs the key file.

Sounds good. Will remove this result and use a default file name in following tasks.

ligangty replied:
Fixed in new commit

mmalina commented on tasks/mrrc-collect-params/mrrc-collect-params.yaml:
```diff
@@ -0,0 +1,60 @@
+          rm $mrrcEnvFile
+        fi
+
+        zipRegistry=$(yq -r '.zipRegistry' $(params.dataJsonPath))
```
Any reason to use `yq` instead of `jq`? I understand that it will likely work, but I'd default to `jq` when working with json files - if nothing else, it makes us consistent.

ligangty replied:
This is because the task is using quay.io/konflux-ci/yq:latest which just provided yq. I don't know what's the official image from konflux which can provide jq.

mmalina replied:
Oh I see, I didn't notice that. What do you think about using our release-service-utils image? It's used in the vast majority of tasks and we plan to update it automatically in the future. It's defined here: https://github.com/konflux-ci/release-service-utils/blob/main/Dockerfile

ligangty replied:
Oh, ok. I'll try to use this image to test.

ligangty replied:
Fixed in new commit

johnbieren commented on tasks/mrrc-collect-params/README.md:
```diff
@@ -0,0 +1,11 @@
+|------|-------------|----------|---------------|
+| dataPath | Path to the merged data JSON file generated by collect-data task and containing the mrrc configuration options to use | No | |
+
+## Changes since 0.1
```
don't need this line

ligangty replied:
Fixed in new commit

johnbieren commented on tasks/mrrc-collect-params/mrrc-collect-params.yaml:
```diff
@@ -0,0 +1,60 @@
+apiVersion: tekton.dev/v1beta1
```
we use tekton.dev/v1

ligangty replied:
Ok will fix.

ligangty replied:
Fixed in new commit

johnbieren commented on tasks/mrrc-collect-params/mrrc-collect-params.yaml:
```diff
@@ -0,0 +1,60 @@
+      description: the secret name for charon aws credential file
+  steps:
+    - name: mrrc-collect-params
+      image: quay.io/konflux-ci/yq:latest
```
can you pin this to a digest? We don't want floating tags in our tasks because then pipelines can break if a bad image is pushed

ligangty replied:
This will be changed to use release-service-utils with a digest tag in new fix.

ligangty replied:
Fixed in new commit

johnbieren commented on tasks/mrrc-collect-params/mrrc-collect-params.yaml:
```diff
@@ -0,0 +1,60 @@
+        awsSecret=$(yq -r '.mrrc.awsSecret' $(params.dataJsonPath))
+        echo -n "$awsSecret" > "$(results.charonAWSSecret.path)"
+
+        consignPubKey=$(yq -r '.cosignPubKeyFile' $(params.dataJsonPath))
```
Shouldn't it be `cosignPubKey` not `consignPubKey`?

ligangty replied:
Oh sorry for the typo. And this value will be removed later.

ligangty replied:
> Why use the snapshotSpec instead of the data file which is already a parameter to this task?

Please see answer here: https://github.com/konflux-ci/release-service-catalog/pull/493#discussion_r1696118803

ligangty replied:
Fixed in new commit

johnbieren commented on tasks/mrrc-collect-params/mrrc-collect-params.yaml:
```diff
@@ -0,0 +1,60 @@
+        echo -n "$awsSecret" > "$(results.charonAWSSecret.path)"
+
+        consignPubKey=$(yq -r '.cosignPubKeyFile' $(params.dataJsonPath))
+        if [ "$consignPubKey" != "" ] && [ "$consignPubKey" != "null" ]; then 
```
If the key is missing, I would think it is (always) either the empty string or null, so we would only have to check that one condition, not both

ligangty replied:
um... I'll check the exact value for the jq empty search and use it later.

ligangty replied:
Seems the jq will return "null" for the empty value, so will just use the null check. Fixed.

mmalina replied:
OK, but I wonder what the intention is here. If it's not set, we will not save this result. I'm not sure right now, but won't it be a problem when you try to use that result value as input for the other task? Wouldn't it be better to just default to ""?

And what about all the other values this script loads? Only this one is checked somehow, but for example if `.mrrc.environment` and `.mrrc.release` are not set, you will end up with target=null-maven-null

johnbieren replied:
If it is a required value, you can add `-e` to the `jq` command so the task will fail if it is missing. We use this elsewhere

ligangty replied:
Actually these values will set in RPA later for non-missing values, but of course, we can add -e to make sure the value will not be missing. I'll add them.

johnbieren replied:
If they are needed, as in whatever task(s) will use the result and will fall over if the value is the empty string, we should add the `-e`

ligangty replied:
Added in fix.

ligangty commented:
I'll try to merge the commits together with sign-off to fix the gitlint checks.

mmalina commented on tasks/collect-mrrc-params/README.md:
```diff
@@ -0,0 +1,10 @@
+# collect-mrrc-params
+
+Tekton task that collects MRRC(maven.repository.redhat.com) configuration options from the data file. MRRC is used to host maven artifacts of RedHat Middleware products.
```
RedHat -> Red Hat

ligangty replied:
Ok. Will fix.

mmalina replied:
It seems this is fixed, thanks.

mmalina commented:
@ligangty please don't resolve threads that you did not open. It makes reviewing harder. Suggestion for next time.

mmalina commented on tasks/collect-mrrc-params/README.md:
I have a general question: What will the main task interact with? Pub? In that case it will most likely need to create an internalRequest (see for example here: https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/push-disk-images/push-disk-images.yaml#L67 ).

This is how we interact with Red Hat services that are not on the public internet - we have a separate cluster that will execute pipelines based on the internalRequest CR that you create in the main pipeline.

The reason I'm asking is that we weren't sure if this collect task is needed at all - all the values from the data file could be loaded in the main task directly. The main reason to have a separate collect task is if you need to mount a secret as a volume - then you need the name of the secret as a task parameter, so a separate collect task is needed for that. (And since you already have a separate task, it makes sense to load all the other values there as well.)

But if we need to run the actual thing via an internalRequest, then the secret will only be loaded in the internalRequest pipeline and so this collect task would not be needed.

ligangty replied:
This task is part of a pure new process which publish MW maven artifacts to MRRC (maven.repository.redhat.com). The MRRC is a public service which will be used to provide maven artifacts to Red Hat customers of MW products. I'm working on another task now which will use a dedicated client tool to publish the content, and it needs several credentials and configurations which should be saved as secrets and configmaps. That's why I'm adding this collect-mrrc-params which includes the mrrc.awsSecret which will be the secrets. And the cosignPubKey will be used to verify the source of the artifacts which is hosted as zip in quay.io. It may be different because there are different MW products which will use this pipeline in future, so the basic idea here is to store them as configmap and let different MW teams to set them in their own RPA.

ligangty replied:
If you are interested in this new process, you can check the jira issue SPMM-16583, which has the doc to describe the whole process.

mmalina replied:
OK, thanks for the details. So it seems that all the services you will interact with are on the public internet and not in Red Hat network, right? In that case it's ok.

ligangty replied:
One thing we're not sure yet is the detached signature support. Current MRRC release process which is executed manually by releng is using rpm-sign which needs to talk to a signing server in the internal network. We're not sure how this will be implemented in the Konflux environment. If it still needs talk to the internal signing server, we should say this will be the only internal service we'll use. 
Some related jira issues are: PVSEC-11984, SIGNSERVER-1358, SIGNSERVER-1086

mmalina replied:
OK. But in that case it will be an additional thing to the main thing which is the push/publish of content to mrrc which is all public. So that still means that the current design looks good (having a collect task).

johnbieren reviewed (commented):
I don't have any other things to change besides what Martin has open. Thanks for addressing the feedback!
Based on the comments it seems this stuff will likely not need an internalrequest, so this task will be needed; which I think is a good thing because the tasks that do not use an internal request are more straight forward

johnbieren commented:
/ok-to-test

ligangty commented:
Hi @johnbieren , I've submitted a new fix for the left issue and some issues for the gitlint and checkton issues. But for the tests failures I'm not sure how to check and fix. Do we have some guide on it?

johnbieren commented on tasks/collect-mrrc-params/collect-mrrc-params.yaml:
```diff
@@ -0,0 +1,50 @@
+        jq -jer '.mrrc.awsSecret' $(params.dataJsonPath) > "$(results.charonAWSSecret.path)"
+
+        cosignPubKey="$(jq -jr '.cosignPubKeyConfig' $(params.dataJsonPath))"
+        if [ "$cosignPubKey" = "null" ]; then 
```
It looks like this is just saying get the value from the data file, if it doesn't exist use `cosign-pub-key`. Have you thought about just using jq's default functionality instead of the if statement? Here is an example (the `//`) https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/apply-mapping/apply-mapping.yaml#L128

ligangty replied:
Ok I'll try this.

ligangty replied:
Added the jq default value in fix.

johnbieren commented on tasks/collect-mrrc-params/tests/test-collect-mrrc-params.yaml:
```diff
@@ -0,0 +1,81 @@
+              
+              test "$(params.charonAWSSecret)" == "charon-aws-credentials"
+
+              . $(params.mrrcParamFilePath)
```
what does this do?

ligangty replied:
The "params.mrrcParamFilePath" is the file which contains the params for mrrc usage like these format:
```shell
export MRRC_REGISTRY=XXX
export MRRC_PRODUCT_NAME=XXX
......
```
so use "." command to set these as environment variables which can be used later in other tasks. (because "source" command will cause checkton warning)

The reason here use this way is to set env vars for following tasks is they're using some image that does not contain "jq" command, which means I can not parse the params what I need from the data.json.

johnbieren replied:
ahh, missed that. Great explanation, thank you

johnbieren commented on tasks/collect-mrrc-params/tests/test-collect-mrrc-params.yaml:
```diff
@@ -0,0 +1,81 @@
+                  "environment": "dev",
+                  "release": "ga"
+                },
+                "cosignPubKeyFile": "test-consign.pub",
```
Any more thought about putting this and the two below it in the mrrc top level key?

ligangty replied:
Oh, sorry this cosignPubKeyFile is not needed anymore and I'll remove it. But the cosignPubKeyConfig is actually not related to mrrc in business level. It is just used to verify the source maven repo zip, so that's why I put it out of the mrrc.

johnbieren replied:
So should we have a maven top level key (like we have mrrc) with this in it? I am just a little wary of putting new top level keys. It is a little confusing to users to have them. I think most if not all of our top level keys have nested keys below them to create a hierarchy style

ligangty replied:
Ok. I'll try to add it.

ligangty replied:
After consideration, I've moved this one under the "mrrc.product" name as it should be bundled to a product. Fixed in new code.

johnbieren commented on tasks/collect-mrrc-params/tests/test-collect-mrrc-params-fail-no-data.yaml:
```diff
@@ -0,0 +1,22 @@
+      taskRef:
+        name: collect-mrrc-params
+      params:
+        - name: dataPath
```
This should be `dataJsonPath` instead of `dataPath` - it is failing your tekton unit tests

ligangty replied:
Oh. I'll fix.

ligangty replied:
Fixed.

johnbieren commented:
> Hi @johnbieren , I've submitted a new fix for the left issue and some issues for the gitlint and checkton issues. But for the tests failures I'm not sure how to check and fix. Do we have some guide on it? And BTW, do we have some offline tools to do these checkton and gitlint check so I can do it locally to avoid submit with unknown check issues?

You failed two things, checkton and unit tests. I will merge without checkton passing - it is a new check for us and the results aren't pasted too clearly just yet. As far as the tekton tests, I am commenting on the files directly

johnbieren commented on tasks/collect-mrrc-params/tests/test-collect-mrrc-params.yaml:
```diff
@@ -0,0 +1,81 @@
+          workspace: tests-workspace
+      runAfter:
+        - setup
+    - name: check-result
```
This task needs a 
```
runAfter:
  - run-task
```

ligangty replied:
Ok. Will fix.

ligangty replied:
Fixed.

ligangty commented:
/ok-to-test

ligangty commented:
Hi @johnbieren I've updated the fix with recent review results. Feel free to check when you are not busy. Thanks.

johnbieren commented on tasks/collect-mrrc-params/tests/test-collect-mrrc-params.yaml:
```diff
@@ -0,0 +1,82 @@
+              test "$MRRC_TARGET" == "dev-maven-ga"
+              test "$MRRC_PRODUCT_NAME" == "activemq"
+              test "$MRRC_PRODUCT_VERSION" == "0.0.1"
+            runAfter:
```
The indentation of this isn't right. Check out how you have it on line 50 of this file

ligangty replied:
oops... Sorry, fixed.

johnbieren commented on tasks/collect-mrrc-params/collect-mrrc-params.yaml:
```diff
@@ -0,0 +1,47 @@
+        printf "$cosignPubKey" > "$(results.cosignPubKeyConfig.path)"
+
+        printf "$MRRC_ENV_FILE_PATH" > "$(results.mrrcParamFilePath.path)"
+      workingDir: $(workspaces.data.path)
```
hmm I don't think we need this line. I don't see it in the other tasks and they do very similar things

ligangty replied:
Removed

johnbieren reviewed (commented):
Two small comments left

johnbieren commented:
Needs rebase too

ligangty commented:
> Needs rebase too

Done

ligangty commented:
@johnbieren I'm a little confused about the "e2e test" and "Tekton tests" result. Can you help point out what are the errors there?

johnbieren commented on tasks/collect-mrrc-params/tests/test-collect-mrrc-params.yaml:
```diff
@@ -0,0 +1,82 @@
+        - name: data
+          workspace: tests-workspace
+      taskSpec:
+        workspaces:
```
This and the following line should not be in the `taskSpec` I think that is the issue

ligangty replied:
Um... interesting. This is what I copied from the https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/collect-pyxis-params/tests/test-collect-pyxis-params.yaml#L16 , but not sure why it failed. Whatever I've removed them.

johnbieren replied:
Yeah, you're right, that wasn't the issue. I'll comment once I find it. You will also need to rebase this

johnbieren commented:
> @johnbieren I'm a little confused about the "e2e test" and "Tekton tests" result. Can you help point out what are the errors there?

4 things failed. 
- tekton unit tests - `test-collect-mrrc-params` was never ready. This usually means some syntax error in your test file. I commented what I think is the problem
- checkton - this is a new check that will fail in most of our tasks, so until we fix it across the board, I am allowing PRs from members outside of our team that fail this check
- e2e suite - You add a new task in this PR that isn't even part of a pipeline. The failure is just from a e2e flake. I can merge bypassing this
- commit isn't signed https://github.com/konflux-ci/release-service/blob/main/CONTRIBUTING.md#signing-commits

ligangty committed:
feat(MMENG-4168): add mrrc-collect-params task

This commit adds a task for collecting params for MRRC from data
json of collect-data task.

Signed-off-by: Gang Li <ligangty@users.noreply.github.com>

ligangty commented:
> > @johnbieren I'm a little confused about the "e2e test" and "Tekton tests" result. Can you help point out what are the errors there?
> 
> 4 things failed.
> 
> * tekton unit tests - `test-collect-mrrc-params` was never ready. This usually means some syntax error in your test file. I commented what I think is the problem

Ok I've removed them.

> * checkton - this is a new check that will fail in most of our tasks, so until we fix it across the board, I am allowing PRs from members outside of our team that fail this check
> * e2e suite - You add a new task in this PR that isn't even part of a pipeline. The failure is just from a e2e flake. I can merge bypassing this
> * commit isn't signed https://github.com/konflux-ci/release-service/blob/main/CONTRIBUTING.md#signing-commits

Ok, I've merged all commits together into one with signature.

johnbieren commented on tasks/collect-mrrc-params/tests/test-collect-mrrc-params.yaml:
```diff
@@ -0,0 +1,80 @@
+          value: $(tasks.run-task.results.cosignPubKeyConfig)
+        - name: charonAWSSecret
+          value: $(tasks.run-task.results.charonAWSSecret)
+      taskSpec:
```
Can you try adding the params here? See https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/collect-pyxis-params/tests/test-collect-pyxis-params.yaml#L52

ligangty replied:
Added. And code rebased.

ligangty commented:
Hi @johnbieren Please help review and test recent change. Thanks!

mmalina commented on tasks/collect-mrrc-params/collect-mrrc-params.yaml:
```diff
@@ -0,0 +1,46 @@
+
+        MRRC_ENV_FILE_PATH="$(dirname $(params.dataJsonPath))/mrrc.env"
+
+        environment="$(jq -jer '.mrrc.environment' $(params.dataJsonPath))"
```
What is the `-j` for? According to jq manual:

> Like -r but jq won´t print a newline after each output.

So that seems to imply you want to use either `-j` or `-r`. There is nothing to join here imo, so I think we only want `-er`. (This applies to all of these `jq` commands.)

ligangty replied:
It's copied from the collect-pyxis-params task: https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/collect-pyxis-params/collect-pyxis-params.yaml#L49

Seems the -r will bring a newline, actually I don't think a newline is needed here, so maybe -j is needed other than -r. I'll fix with -j instead of -r

ligangty replied:
Fixed.

mmalina replied:
> It's copied from the collect-pyxis-params task

I wasn't aware it was used there. But yeah, that seems wrong :)

I would preferably go with `-r` as that's what we use in most places and should be sufficient (the newline doesn't matter - it will get stripped before it's assigned to the variable). But `-j` seems to be ok too.

mmalina reviewed (commented):
I think it should be mostly ready to be merged now, but there are some failed checks.

The tekton task tests fail. Let me know if you need help with that.

Checkton check also has some issues - please see the "Run Checkton" step here: https://github.com/konflux-ci/release-service-catalog/actions/runs/10327134604/job/28650181171?pr=493

For the `printf` issue - you could also just use `echo -n`.

ligangty commented:
> I think it should be mostly ready to be merged now, but there are some failed checks.
> 
> The tekton task tests fail. Let me know if you need help with that.
> 
> Checkton check also has some issues - please see the "Run Checkton" step here: https://github.com/konflux-ci/release-service-catalog/actions/runs/10327134604/job/28650181171?pr=493
> 
> For the `printf` issue - you could also just use `echo -n`.

Actually I used `echo -n` in the original post, but I got error in checkton check which said `echo -n` is not valid in /bin/sh env, and suggest using printf with "\n" instead. So I'm a little confused in terms of these checkton checks.

mmalina commented:
> > For the `printf` issue - you could also just use `echo -n`.
> 
> Actually I used `echo -n` in the original post, but I got error in checkton check which said `echo -n` is not valid in /bin/sh env, and suggest using printf with "\n" instead. So I'm a little confused in terms of these checkton checks.

Haha, ok, I didn't know that :) The task could actually use a shebang. Otherwise `#!/bin/sh` is implied: https://tekton.dev/docs/pipelines/tasks/#running-scripts-within-steps Not sure if Checkton makes the right assumption here. But we had many places for example with here-variables (`<<<`) which are not really supported in POSIX shell, so we switched to explicitly use `bash` in many places. I would recommend you use the same: `#!/usr/bin/env bash`

Hopefully with that Checkton will no longer have an issue with `echo -n`.

ligangty commented:
> > > For the `printf` issue - you could also just use `echo -n`.
> > 
> > 
> > Actually I used `echo -n` in the original post, but I got error in checkton check which said `echo -n` is not valid in /bin/sh env, and suggest using printf with "\n" instead. So I'm a little confused in terms of these checkton checks.
> 
> Haha, ok, I didn't know that :) The task could actually use a shebang. Otherwise `#!/bin/sh` is implied: https://tekton.dev/docs/pipelines/tasks/#running-scripts-within-steps Not sure if Checkton makes the right assumption here. But we had many places for example with here-variables (`<<<`) which are not really supported in POSIX shell, so we switched to explicitly use `bash` in many places. I would recommend you use the same: `#!/usr/bin/env bash`
> 
> Hopefully with that Checkton will no longer have an issue with `echo -n`.

Ok. I'll change back to "echo -n" with "#!/usr/bin/env bash". Hope it will not complain. :)

BTW, I've changed the above comment for the "tekton tests". Sorry for that.

johnbieren commented:
> > > For the `printf` issue - you could also just use `echo -n`.
> > 
> > 
> > Actually I used `echo -n` in the original post, but I got error in checkton check which said `echo -n` is not valid in /bin/sh env, and suggest using printf with "\n" instead. So I'm a little confused in terms of these checkton checks.
> 
> Haha, ok, I didn't know that :) The task could actually use a shebang. Otherwise `#!/bin/sh` is implied: https://tekton.dev/docs/pipelines/tasks/#running-scripts-within-steps Not sure if Checkton makes the right assumption here. But we had many places for example with here-variables (`<<<`) which are not really supported in POSIX shell, so we switched to explicitly use `bash` in many places. I would recommend you use the same: `#!/usr/bin/env bash`
> 
> Hopefully with that Checkton will no longer have an issue with `echo -n`.

I am bypassing failed checkton results until we have completed the epic to fix all checkton issues in our tasks. I don't think it is fair to external contributors to enforce it yet, as copying from an existing task is a great idea but the existing tasks are all full of checkton issues for now

johnbieren commented on tasks/collect-mrrc-params/collect-mrrc-params.yaml:
```diff
@@ -0,0 +1,47 @@
+
+        MRRC_ENV_FILE_PATH="$(dirname $(params.dataJsonPath))/mrrc.env"
+
+        environment="$(jq -re '.mrrc.environment' $(params.dataJsonPath))"
```
Ah (this is true for all of your jq statements). I think you want `$(workspaces.data.path)/$(params.dataPath)`. Something like https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/push-snapshot/push-snapshot.yaml#L82 would work too
The tests are failing due to this

mmalina replied:
Heh, you beat me to this :)

mmalina commented:
Fair point about Checkton from Johnny - I kind of forgot about that. So consider fixing those errors optional.

> And for tekton tests, I think I really needs some help on it. The two cases are modified based on the "collect-pyxis-params" tests too, but I'm not quite sure how these tests are running.
> 
> BTW, do we have some ways to run these tests in local env so I can run the tests and fix the issues before I push them for testing?

The task testing is described here (including how to set up local env with `kind`): https://github.com/konflux-ci/release-service-catalog/blob/development/CONTRIBUTING.md#running-tekton-task-tests-manually

What it's failing on now is a missing data file when running the task. Your test looks ok, but it seems to be a problem in the task - I opened a thread in the code.

mmalina commented on tasks/collect-mrrc-params/collect-mrrc-params.yaml:
```diff
@@ -0,0 +1,47 @@
+
+        MRRC_ENV_FILE_PATH="$(dirname $(params.dataJsonPath))/mrrc.env"
+
+        environment="$(jq -re '.mrrc.environment' $(params.dataJsonPath))"
```
`$(params.dataJsonPath))` is the relative path to the file within the workspace. So you will need `"$(workspaces.data.path)/$(params.dataPath)"`. Maybe it's worth setting up a var for that: `DATA_FILE="$(workspaces.data.path)/$(params.dataPath)"`

This applies to other places as well.

ligangty replied:
Ok. Changed.

ligangty commented:
Oh sorry. There is a var set error. I just fixed and submitted.

johnbieren commented on tasks/collect-mrrc-params/collect-mrrc-params.yaml:
```diff
@@ -0,0 +1,50 @@
+
+        DATA_FILE="$(workspaces.data.path)/$(params.dataJsonPath)"
+
+        MRRC_ENV_FILE_PATH="$(dirname $DATA_FILE)/mrrc.env"
```
This also should be in the workspace

mmalina replied:
The way it's implemented now it **is** in the workspace (because it's derived from `$DATA_FILE`. But there is one issue - the result should contain the relative path under the workspace, not the absolute path.

johnbieren replied:
Ah, thanks for correcting me. Hopefully CI passes once this last part is fixed

ligangty replied:
Um... I think the `dirname $DATA_FILE` will give the absolute path of the data.json, including both the workspace dir and the relative dir. Not sure why it can not work in test. But FWIW I've changed the code to use `dirname $(params.dataJsonPath)` here which just use the relative path now, which means this MRRC_ENV_FILE_PATH will just be a relative path in the workspace dir. And I also changed the test according to this change.
